File: api/src/routes/academic.ts
import { Router, Request, Response } from "express";
import prisma from "../lib/prisma.js";
import { z } from "zod";

const router = Router();

const gradePoints: Record<string, number> = {
	"A": 4.0, "A-": 3.7,
	"B+": 3.3, "B": 3.0, "B-": 2.7,
	"C+": 2.3, "C": 2.0, "C-": 1.7,
	"D+": 1.3, "D": 1.0, "F": 0.0
};

const getStudentId = async (userId: number, role: string, targetStudentId?: string) => {
	const userRole = role.toUpperCase();
	if (userRole === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId } });
		return student?.id;
	}
	if ((userRole === "ADMIN" || userRole === "COUNSELOR") && targetStudentId) {
		return parseInt(targetStudentId);
	}
	return null;
};

router.get("/", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.query.studentId as string);
	console.log(`[ACADEMIC GET] User: ${user.email}, Role: ${user.role}, Requested StudentId: ${req.query.studentId}, Resolved StudentId: ${studentId}`);
	
	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const records = await prisma.academicRecord.findMany({
		where: { studentId },
		orderBy: [{ year: "desc" }, { semester: "desc" }]
	});

	console.log(`[ACADEMIC GET] Found ${records.length} records`);
	res.json(records);
});

const academicSchema = z.object({
	courseName: z.string(),
	grade: z.string().optional(),
	credits: z.number().optional(),
	semester: z.string(),
	year: z.number(),
	isAP: z.boolean().default(false),
	isHonors: z.boolean().default(false),
	studentId: z.number().optional()
});

router.post("/", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const result = academicSchema.safeParse(req.body);
	if (!result.success) return res.status(400).json({ error: result.error.format() });

	let studentId = result.data.studentId;
	if (user.role.toUpperCase() === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId: user.id } });
		if (!student) return res.status(404).json({ error: "Student record not found" });
		studentId = student.id;
	} else if (!studentId) {
		return res.status(400).json({ error: "studentId is required for non-students" });
	}

	const record = await prisma.academicRecord.create({
		data: {
			...result.data,
			studentId: studentId!
		}
	});

	res.status(201).json(record);
});

router.delete("/:id", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const id = parseInt(req.params.id);
	const record = await prisma.academicRecord.findUnique({ where: { id } });
	if (!record) return res.status(404).json({ error: "Record not found" });

	if (user.role.toUpperCase() === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId: user.id } });
		if (!student || record.studentId !== student.id) {
			return res.status(403).json({ error: "Forbidden" });
		}
	}

	await prisma.academicRecord.delete({ where: { id } });
	res.status(204).end();
});

router.get("/report-data", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.query.studentId as string);
	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const records = await prisma.academicRecord.findMany({
		where: { studentId },
		orderBy: [{ year: "asc" }, { semester: "asc" }]
	});

	// Transform to trend data
	const trend = records.map(r => ({
		semester: `${r.semester} ${r.year}`,
		gpa: gradePoints[r.grade || ""] || 0,
		course: r.courseName
	}));

	res.json(trend);
});

router.get("/gpa", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.query.studentId as string);
	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const records = await prisma.academicRecord.findMany({
		where: { studentId }
	});

	let hypothetical: any[] = [];
	try {
		if (req.query.hypothetical) {
			hypothetical = JSON.parse(req.query.hypothetical as string);
		}
	} catch (e) {}

	const allRecords = [...records, ...hypothetical];

	const calculateGPA = (recs: any[]) => {
		let totalPoints = 0;
		let totalCredits = 0;
		recs.forEach(r => {
			const base = gradePoints[r.grade || ""] || 0;
			const weight = r.isAP ? 1.0 : (r.isHonors ? 0.5 : 0.0);
			const credits = r.credits || 1;
			totalPoints += (base + weight) * credits;
			totalCredits += credits;
		});
		return totalCredits > 0 ? totalPoints / totalCredits : 0;
	};

	const currentGPA = calculateGPA(records);
	const potentialGPA = calculateGPA(allRecords);

	res.json({ 
		currentGPA: parseFloat(currentGPA.toFixed(2)), 
		potentialGPA: parseFloat(potentialGPA.toFixed(2)) 
	});
});

router.post("/generate-report", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.body.studentId);
	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const config = await prisma.systemConfig.findFirst();
	const agentConfig = config?.agentConfig as any;
	if (!agentConfig?.report?.enabled) {
		return res.status(400).json({ error: "Report Generator is disabled" });
	}

	res.json({ message: "Report agent triggered", studentId });
});

export default router;

File: api_py/routers/academic.py
from fastapi import APIRouter, Depends, HTTPException, Request
from ..database import execute_query, execute_commit
from ..auth import get_current_user
from ..utils import calculate_weighted_gpa
from pydantic import BaseModel
from typing import Optional
import traceback

router = APIRouter()

class AcademicRecord(BaseModel):
    courseName: str
    grade: Optional[str] = "A"
    credits: float = 4.0
    semester: str = "Fall"
    year: int = 2023
    isAP: bool = False
    isHonors: bool = False
    studentId: Optional[int] = None

def get_target_student_id(user, requested_student_id):
    try:
        role = user["role"].upper()
        if role == "STUDENT":
            student = execute_query("SELECT id FROM Student WHERE userId = %s", (user["id"],), fetch_one=True)
            return student["id"] if student else None
        
        if role == "PARENT":
            if requested_student_id:
                link = execute_query("SELECT studentId FROM StudentParent WHERE studentId = %s AND parentId = %s", (requested_student_id, user["id"]), fetch_one=True)
                return int(requested_student_id) if link else None
            student = execute_query("SELECT studentId FROM StudentParent WHERE parentId = %s LIMIT 1", (user["id"],), fetch_one=True)
            return student["studentId"] if student else None

        if (role in ["ADMIN", "COUNSELOR"]) and requested_student_id:
            return int(requested_student_id)
        return None
    except Exception as e:
        print(f"[ACADEMIC] Error in get_target_student_id: {e}")
        return None

@router.get("/")
def get_academics(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, studentId)
        print(f"[ACADEMIC] GET / - User: {current_user['id']}, Role: {current_user['role']}, Target ID: {target_id}")
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
        
        records = execute_query("SELECT * FROM AcademicRecord WHERE studentId = %s ORDER BY year DESC, semester DESC", (target_id,))
        return records
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in GET /: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/")
def add_academic(record: AcademicRecord, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, record.studentId)
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
        
        query = """
            INSERT INTO AcademicRecord (studentId, courseName, grade, credits, semester, year, isAP, isHonors)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
        """
        params = (target_id, record.courseName, record.grade, record.credits, record.semester, record.year, record.isAP, record.isHonors)
        record_id = execute_commit(query, params)
        return {"id": record_id, **record.dict()}
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in POST /: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{id}/")
def delete_academic(id: int, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, None)
        if current_user["role"].upper() == "STUDENT":
            if not target_id:
                raise HTTPException(status_code=404, detail="Student not found")
            record = execute_query("SELECT studentId FROM AcademicRecord WHERE id = %s", (id,), fetch_one=True)
            if not record or record["studentId"] != target_id:
                raise HTTPException(status_code=403, detail="Forbidden")
                
        execute_commit("DELETE FROM AcademicRecord WHERE id = %s", (id,))
        return {"success": True}
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in DELETE /{id}: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/gpa/")
def get_gpa(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, studentId)
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
            
        records = execute_query("SELECT * FROM AcademicRecord WHERE studentId = %s", (target_id,))
        current_gpa = calculate_weighted_gpa(records)
        return {"currentGPA": current_gpa, "potentialGPA": current_gpa}
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in GET /gpa/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/report-data/")
def get_report_data(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, studentId)
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
            
        records = execute_query("SELECT * FROM AcademicRecord WHERE studentId = %s ORDER BY year ASC, semester ASC", (target_id,))
        
        grade_points = {
            "A": 4.0, "A-": 3.7, "B+": 3.3, "B": 3.0, "B-": 2.7,
            "C+": 2.3, "C": 2.0, "C-": 1.7, "D+": 1.3, "D": 1.0, "F": 0.0
        }
        
        trend = []
        for r in records:
            base = grade_points.get(r["grade"] or "", 0.0)
            weight = 1.0 if r["isAP"] else (0.5 if r["isHonors"] else 0.0)
            trend.append({
                "semester": f"{r['semester']} {r['year']}",
                "gpa": base + weight,
                "course": r["courseName"]
            })
        return trend
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in GET /report-data/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/readiness/")
def get_student_readiness(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, studentId)
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
            
        student = execute_query("SELECT grade FROM Student WHERE id = %s", (target_id,), fetch_one=True)
        if not student:
             raise HTTPException(status_code=404, detail="Student not found")
             
        records = execute_query("SELECT grade, credits, isAP, isHonors FROM AcademicRecord WHERE studentId = %s", (target_id,))
        current_gpa = calculate_weighted_gpa(records)
        
        grade = student["grade"]
        milestones = []
        if grade == 9:
            milestones = [{"task": "Join 2 ECs", "status": "Done"}, {"task": "Course Selection", "status": "Todo"}]
        elif grade == 10:
            milestones = [{"task": "PSAT Prep", "status": "Done"}, {"task": "Summer Internships", "status": "Todo"}]
        elif grade == 11:
            milestones = [{"task": "SAT/ACT", "status": "Todo"}, {"task": "College List", "status": "Todo"}]
        else:
            milestones = [{"task": "Apply", "status": "Todo"}, {"task": "Scholarships", "status": "Todo"}]
            
        benchmarks = [
            {"college": "Reach (Ivy)", "medianGpa": 3.9, "diff": round(current_gpa - 3.9, 2)},
            {"college": "Match (State)", "medianGpa": 3.5, "diff": round(current_gpa - 3.5, 2)},
            {"college": "Safety", "medianGpa": 3.0, "diff": round(current_gpa - 3.0, 2)}
        ]
        
        readiness = (current_gpa / 4.0 * 60) + (min(len(records), 10) * 4)
        
        return {
            "currentGpa": current_gpa,
            "milestones": milestones,
            "benchmarks": benchmarks,
            "readinessScore": round(min(readiness, 100))
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ACADEMIC] Exception in GET /readiness/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))
